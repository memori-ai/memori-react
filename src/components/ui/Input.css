.memori-input-wrapper {
    width: 100%;
    margin-bottom: 1rem;
    font-family: inherit;
  }
  
  .memori-input--label {
    display: block;
    margin-bottom: 0.5rem;
    color: #333;
    font-weight: 500;
  }
  
  .memori-input-container {
    display: flex;
    overflow: hidden;
    width: 100%;
    border-radius: 0.25rem;
  }
  
  .memori-input--field-container {
    position: relative;
    display: flex;
    flex: 1;
    align-items: center;
    border: 1px solid #d9d9d9;
    border-radius: 0.25rem;
    background-color: #fff;
    transition: all 0.3s;
  }
  
  .memori-input--with-addon-before .memori-input--field-container {
    border-bottom-left-radius: 0;
    border-top-left-radius: 0;
  }
  
  .memori-input--with-addon-after .memori-input--field-container {
    border-bottom-right-radius: 0;
    border-top-right-radius: 0;
  }
  
  .memori-input--field {
    width: 100%;
    padding: 0.5rem 0.75rem;
    border: none;
    background: transparent;
    color: #333;
    font-size: 14px;
    outline: none;
  }
  
  .memori-input--field::placeholder {
    color: #bfbfbf;
  }

  .memori-input--disabled .memori-input--field-container {
    border-color: #d9d9d9;
    background-color: #f5f5f5;
    color: rgba(0, 0, 0, 0.25);
    cursor: not-allowed;
  }
  
  
  .memori-input--field-container:hover:not(.memori-input--disabled):not(.memori-input--readonly) {
    border-color: #40a9ff;
  }
  
  .memori-input--field-container:focus-within:not(.memori-input--disabled):not(.memori-input--readonly) {
    border-color: #40a9ff;
    box-shadow: 0 0 0 2px rgba(24, 144, 255, 0.2);
  }
  
  .memori-input--disabled .memori-input--field,
  .memori-input--readonly .memori-input--field {
    background-color: #f5f5f5;
    color: rgba(0, 0, 0, 0.25);
    cursor: not-allowed;
  }
  
  .memori-input--addon-before,
  .memori-input--addon-after {
    display: flex;
    align-items: center;
    padding: 0 0.75rem;
    border: 1px solid #d9d9d9;
    background-color: #fafafa;
    white-space: nowrap;
  }
  
  .memori-input--addon-before {
    border-right: none;
    border-bottom-left-radius: 0.25rem;
    border-top-left-radius: 0.25rem;
  }
  
  .memori-input--addon-after {
    border-left: none;
    border-bottom-right-radius: 0.25rem;
    border-top-right-radius: 0.25rem;
  }
  
  .memori-input--prefix,
  .memori-input--suffix,
  .memori-input--clear {
    display: flex;
    align-items: center;
    justify-content: center;
    color: rgba(0, 0, 0, 0.45);
  }
  
  .memori-input--prefix {
    margin-left: 0.75rem;
  }
  
  .memori-input--suffix,
  .memori-input--clear {
    margin-right: 0.75rem;
  }
  
  .memori-input--clear {
    width: 1rem;
    height: 1rem;
    color: #bfbfbf;
    cursor: pointer;
    transition: color 0.3s;
  }

  .memori-input--clear-icon {
    color: #bfbfbf;
  }
  
  .memori-input--clear:hover {
    color: rgba(0, 0, 0, 0.45);
  }
  
  .memori-input--with-prefix .memori-input--field {
    padding-left: 0;
  }
  
  .memori-input--with-suffix .memori-input--field {
    padding-right: 0;
  }