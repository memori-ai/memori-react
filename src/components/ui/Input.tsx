import React, { ReactNode, useState, useRef, useEffect, Fragment } from 'react';
import cx from 'classnames';
import { Transition } from '@headlessui/react';
import CloseIcon from '../icons/Close';

export interface Props {
  className?: string;
  type?: 'text' | 'password' | 'email' | 'number' | 'search' | 'tel' | 'url';
  name?: string;
  value?: string;
  defaultValue?: string;
  onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void;
  onPressEnter?: (e: React.KeyboardEvent<HTMLInputElement>) => void;
  placeholder?: string;
  disabled?: boolean;
  readOnly?: boolean;
  allowClear?: boolean;
  addonBefore?: ReactNode;
  addonAfter?: ReactNode;
  prefix?: ReactNode;
  suffix?: ReactNode;
  label?: string;
  id?: string;
  maxLength?: number;
  autoFocus?: boolean;
  // Added search related props to match Ant Design
  showSearch?: boolean;
  onSearch?: (value: string) => void;
  // Help text
  help?: ReactNode;
  // Validation status
  status?: 'error' | 'warning' | 'success';
}

const Input = ({
  className,
  type = 'text',
  name,
  value,
  defaultValue,
  onChange,
  onPressEnter,
  placeholder,
  disabled = false,
  readOnly = false,
  allowClear = false,
  addonBefore,
  addonAfter,
  prefix,
  suffix,
  label,
  id,
  maxLength,
  autoFocus = false,
  showSearch = false,
  onSearch,
  help,
  status,
}: Props) => {
  const [innerValue, setInnerValue] = useState(defaultValue || '');
  const [isFocused, setIsFocused] = useState(false);
  const inputRef = useRef<HTMLInputElement>(null);
  
  // Determine if controlled or uncontrolled
  const isControlled = value !== undefined;
  const inputValue = isControlled ? value : innerValue;
  
  // Sync with external value changes
  useEffect(() => {
    if (isControlled && value !== undefined) {
      // If component is controlled, update innerValue to match value prop
      setInnerValue(value);
    }
  }, [isControlled, value]);
  
  // Focus management
  useEffect(() => {
    if (autoFocus && inputRef.current) {
      inputRef.current.focus();
    }
  }, [autoFocus]);
  
  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const newValue = e.target.value;
    
    if (!isControlled) {
      setInnerValue(newValue);
    }
    
    if (onChange) {
      onChange(e);
    }
  };
  
  const handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (e.key === 'Enter') {
      if (showSearch && onSearch) {
        onSearch(inputValue);
      }
      
      if (onPressEnter) {
        onPressEnter(e);
      }
    }
  };
  
  const handleClear = () => {
    // Create a synthetic event that mimics a change event
    const syntheticEvent = {
      target: {
        name,
        value: ''
      },
      currentTarget: {
        name,
        value: ''
      },
      preventDefault: () => {},
      stopPropagation: () => {}
    } as React.ChangeEvent<HTMLInputElement>;
    
    if (!isControlled) {
      setInnerValue('');
    }
    
    if (onChange) {
      onChange(syntheticEvent);
    }
    
    // Focus input after clearing
    if (inputRef.current) {
      inputRef.current.focus();
    }
  };
  
  const showClearButton = allowClear && inputValue && !disabled && !readOnly;
  
  return (
    <div className={cx('memori-input-wrapper', className)}>
      {label && (
        <label htmlFor={id || name} className="memori-input--label">
          {label}
        </label>
      )}
      
      <div className={cx('memori-input-container', {
        'memori-input--disabled': disabled,
        'memori-input--readonly': readOnly,
        'memori-input--with-addon-before': addonBefore,
        'memori-input--with-addon-after': addonAfter,
        'memori-input--with-prefix': prefix,
        'memori-input--with-suffix': suffix || showClearButton,
        'memori-input--focused': isFocused,
        'memori-input--error': status === 'error',
        'memori-input--warning': status === 'warning',
        'memori-input--success': status === 'success'
      })}>
        {addonBefore && (
          <div className="memori-input--addon-before">
            {addonBefore}
          </div>
        )}
        
        <div className={cx('memori-input--field-container', {
          'memori-input--field-focused': isFocused,
          'memori-input--field-error': status === 'error',
          'memori-input--field-warning': status === 'warning',
          'memori-input--field-success': status === 'success'
        })}>
          {prefix && (
            <div className="memori-input--prefix">
              {prefix}
            </div>
          )}
          
          <input
            ref={inputRef}
            className="memori-input--field"
            type={type}
            id={id || name}
            name={name}
            value={inputValue}
            onChange={handleChange}
            onKeyDown={handleKeyDown}
            onFocus={() => setIsFocused(true)}
            onBlur={() => setIsFocused(false)}
            placeholder={placeholder}
            disabled={disabled}
            readOnly={readOnly}
            maxLength={maxLength}
            autoFocus={autoFocus}
            role={type === 'search' ? 'searchbox' : undefined}
          />
          
          <Transition
            as={Fragment}
            show={showClearButton || false}
            enter="transition ease-out duration-100"
            enterFrom="opacity-0 scale-95"
            enterTo="opacity-100 scale-100"
            leave="transition ease-in duration-75"
            leaveFrom="opacity-100 scale-100"
            leaveTo="opacity-0 scale-95"
          >
            <div 
              className="memori-input--clear" 
              onClick={handleClear}
              title="Clear"
              role="button"
              tabIndex={0}
              onKeyDown={(e) => {
                if (e.key === 'Enter' || e.key === ' ') {
                  handleClear();
                  e.preventDefault();
                }
              }}
            >
              <CloseIcon className="memori-input--clear-icon" />
            </div>
          </Transition>
          
          {suffix && !showClearButton && (
            <div className="memori-input--suffix">
              {suffix}
            </div>
          )}
        </div>
        
        {addonAfter && (
          <div className="memori-input--addon-after">
            {addonAfter}
          </div>
        )}
      </div>
      
      {help && (
        <Transition
          show={!!help}
          enter="transition ease-out duration-100"
          enterFrom="opacity-0 -translate-y-1"
          enterTo="opacity-100 translate-y-0"
          leave="transition ease-in duration-75"
          leaveFrom="opacity-100 translate-y-0"
          leaveTo="opacity-0 -translate-y-1"
        >
          <div className={cx('memori-input--help', {
            'memori-input--help-error': status === 'error',
            'memori-input--help-warning': status === 'warning',
            'memori-input--help-success': status === 'success'
          })}>
            {help}
          </div>
        </Transition>
      )}
    </div>
  );
};

export default Input;