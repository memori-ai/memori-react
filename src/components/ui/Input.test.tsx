import React from 'react';
import { render, screen } from '@testing-library/react';
import Input from './Input';

describe('Input Component', () => {
  it('renders Input unchanged', () => {
    const { container } = render(
      <Input
        name="test-input"
        onChange={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders Input with label unchanged', () => {
    const { container } = render(
      <Input
        label="Input Label"
        name="test-input"
        onChange={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders Input with placeholder unchanged', () => {
    const { container } = render(
      <Input
        placeholder="Enter text here"
        name="test-input"
        onChange={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders Input with value unchanged', () => {
    const { container } = render(
      <Input
        value="Test Value"
        name="test-input"
        onChange={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders Input with prefix unchanged', () => {
    const { container } = render(
      <Input
        prefix={<span data-testid="prefix-icon">@</span>}
        name="test-input"
        onChange={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders Input with suffix unchanged', () => {
    const { container } = render(
      <Input
        suffix={<span data-testid="suffix-icon">%</span>}
        name="test-input"
        onChange={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders Input with addonBefore unchanged', () => {
    const { container } = render(
      <Input
        addonBefore="https://"
        name="test-input"
        onChange={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders Input with addonAfter unchanged', () => {
    const { container } = render(
      <Input
        addonAfter=".com"
        name="test-input"
        onChange={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders Input with allowClear unchanged', () => {
    const { container } = render(
      <Input
        value="Clear me"
        allowClear
        name="test-input"
        onChange={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders disabled Input unchanged', () => {
    const { container } = render(
      <Input
        value="Disabled input"
        disabled
        name="test-input"
        onChange={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders password Input unchanged', () => {
    const { container } = render(
      <Input
        type="password"
        value="password123"
        name="test-input"
        onChange={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders search Input unchanged', () => {
    const { container } = render(
      <Input
        type="search"
        placeholder="Search..."
        name="test-input"
        onChange={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders complete example unchanged', () => {
    const { container } = render(
      <Input
        label="Search"
        type="search"
        name="query"
        allowClear
        placeholder="Search by name"
        prefix={<span>🔍</span>}
        addonAfter={<button>Search</button>}
        onChange={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });
});