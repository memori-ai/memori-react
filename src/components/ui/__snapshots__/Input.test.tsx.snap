// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Component renders Input unchanged 1`] = `
<div>
  <div
    class="memori-input-wrapper"
  >
    <div
      class="memori-input-container"
    >
      <div
        class="memori-input--field-container"
      >
        <input
          class="memori-input--field"
          id="test-input"
          name="test-input"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Input Component renders Input with addonAfter unchanged 1`] = `
<div>
  <div
    class="memori-input-wrapper"
  >
    <div
      class="memori-input-container memori-input--with-addon-after"
    >
      <div
        class="memori-input--field-container"
      >
        <input
          class="memori-input--field"
          id="test-input"
          name="test-input"
          type="text"
          value=""
        />
      </div>
      <div
        class="memori-input--addon-after"
      >
        .com
      </div>
    </div>
  </div>
</div>
`;

exports[`Input Component renders Input with addonBefore unchanged 1`] = `
<div>
  <div
    class="memori-input-wrapper"
  >
    <div
      class="memori-input-container memori-input--with-addon-before"
    >
      <div
        class="memori-input--addon-before"
      >
        https://
      </div>
      <div
        class="memori-input--field-container"
      >
        <input
          class="memori-input--field"
          id="test-input"
          name="test-input"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Input Component renders Input with allowClear unchanged 1`] = `
<div>
  <div
    class="memori-input-wrapper"
  >
    <div
      class="memori-input-container memori-input--with-suffix"
    >
      <div
        class="memori-input--field-container"
      >
        <input
          class="memori-input--field"
          id="test-input"
          name="test-input"
          type="text"
          value="Clear me"
        />
        <div
          class="memori-input--clear"
          role="button"
          tabindex="0"
          title="Clear"
        >
          <svg
            aria-hidden="true"
            class="memori-input--clear-icon"
            focusable="false"
            role="img"
            viewBox="0 0 1024 1024"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 0 0 203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Input Component renders Input with label unchanged 1`] = `
<div>
  <div
    class="memori-input-wrapper"
  >
    <label
      class="memori-input--label"
      for="test-input"
    >
      Input Label
    </label>
    <div
      class="memori-input-container"
    >
      <div
        class="memori-input--field-container"
      >
        <input
          class="memori-input--field"
          id="test-input"
          name="test-input"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Input Component renders Input with placeholder unchanged 1`] = `
<div>
  <div
    class="memori-input-wrapper"
  >
    <div
      class="memori-input-container"
    >
      <div
        class="memori-input--field-container"
      >
        <input
          class="memori-input--field"
          id="test-input"
          name="test-input"
          placeholder="Enter text here"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Input Component renders Input with prefix unchanged 1`] = `
<div>
  <div
    class="memori-input-wrapper"
  >
    <div
      class="memori-input-container memori-input--with-prefix"
    >
      <div
        class="memori-input--field-container"
      >
        <div
          class="memori-input--prefix"
        >
          <span
            data-testid="prefix-icon"
          >
            @
          </span>
        </div>
        <input
          class="memori-input--field"
          id="test-input"
          name="test-input"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Input Component renders Input with suffix unchanged 1`] = `
<div>
  <div
    class="memori-input-wrapper"
  >
    <div
      class="memori-input-container memori-input--with-suffix"
    >
      <div
        class="memori-input--field-container"
      >
        <input
          class="memori-input--field"
          id="test-input"
          name="test-input"
          type="text"
          value=""
        />
        <div
          class="memori-input--suffix"
        >
          <span
            data-testid="suffix-icon"
          >
            %
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Input Component renders Input with value unchanged 1`] = `
<div>
  <div
    class="memori-input-wrapper"
  >
    <div
      class="memori-input-container"
    >
      <div
        class="memori-input--field-container"
      >
        <input
          class="memori-input--field"
          id="test-input"
          name="test-input"
          type="text"
          value="Test Value"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Input Component renders complete example unchanged 1`] = `
<div>
  <div
    class="memori-input-wrapper"
  >
    <label
      class="memori-input--label"
      for="query"
    >
      Search
    </label>
    <div
      class="memori-input-container memori-input--with-addon-after memori-input--with-prefix"
    >
      <div
        class="memori-input--field-container"
      >
        <div
          class="memori-input--prefix"
        >
          <span>
            üîç
          </span>
        </div>
        <input
          class="memori-input--field"
          id="query"
          name="query"
          placeholder="Search by name"
          role="searchbox"
          type="search"
          value=""
        />
      </div>
      <div
        class="memori-input--addon-after"
      >
        <button>
          Search
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Input Component renders disabled Input unchanged 1`] = `
<div>
  <div
    class="memori-input-wrapper"
  >
    <div
      class="memori-input-container memori-input--disabled"
    >
      <div
        class="memori-input--field-container"
      >
        <input
          class="memori-input--field"
          disabled=""
          id="test-input"
          name="test-input"
          type="text"
          value="Disabled input"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Input Component renders password Input unchanged 1`] = `
<div>
  <div
    class="memori-input-wrapper"
  >
    <div
      class="memori-input-container"
    >
      <div
        class="memori-input--field-container"
      >
        <input
          class="memori-input--field"
          id="test-input"
          name="test-input"
          type="password"
          value="password123"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Input Component renders search Input unchanged 1`] = `
<div>
  <div
    class="memori-input-wrapper"
  >
    <div
      class="memori-input-container"
    >
      <div
        class="memori-input--field-container"
      >
        <input
          class="memori-input--field"
          id="test-input"
          name="test-input"
          placeholder="Search..."
          role="searchbox"
          type="search"
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;
